Add map-search directive

asaunier opened the pull request:
Fixes https://github.com/c2corg/v6_ui/issues/304

Relies on the geonames service.

tsauerwein commented on c2corg_ui/static/js/map/search.js:
```diff
@@ -0,0 +1,165 @@
+    bindToController: true,
+    controllerAs: 'searchCtrl',
+    template:
+        '<input type="text" placeholder="{{\'Recenter on...\' | translate}}" ' +
```
Is "Recenter on" extracted correctly?

asaunier replied:
I think so.

tsauerwein commented:
:+1:

asaunier commented:
Thanks for reviewing and for the original code @tsauerwein 
Actually the PR is not ready (WIP): the request to geonames is triggered and I can see a valid response in my console but no results are displayed in the dropdown.
Perhaps I should specify somewhere how to read the response? Like here: https://github.com/camptocamp/cgxp/blob/master/core/src/script/CGXP/plugins/Geonames.js#L103-L115 ?

tsauerwein commented:
> Perhaps I should specify somewhere how to read the response?

`CreateGeoJSONBloodhound` expects GeoJSON. You will have convert the response similar to what is done for the simple search: https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/search/simplesearch.js#L299

asaunier commented:
I see. The geonames response is not at all a GeoJSON response:
> {"totalResultsCount":5,"geonames":[{"adminCode1":"24","lng":"1.44006","geonameId":3038744,"toponymName":"Abondant","countryId":"3017382","fcl":"P","population":1807,"countryCode":"FR","name":"Abondant","fclName":"city, village,...","countryName":"France","fcodeName":"populated place","adminName1":"Centre","lat":"48.7859","fcode":"PPL"},{"adminCode1":"25","lng":"21.5159","geonameId":245752,"toponymName":"Abonday","countryId":"2434508","fcl":"P","population":0,"countryCode":"TD","name":"Abonday","fclName":"city, village,...","countryName":"Tchad","fcodeName":"populated place","adminName1":"Région du Sila","lat":"11.8618","fcode":"PPL"},{"adminCode1":"84","lng":"6.72026","geonameId":3038745,"toponymName":"Abondance","countryId":"3017382","fcl":"P","population":1498,"countryCode":"FR","name":"Abondance","fclName":"city, village,...","countryName":"France","fcodeName":"populated place","adminName1":"Auvergne-Rhône-Alpes","lat":"46.28031","fcode":"PPL"},{"adminCode1":"MQ","lng":"-60.88308","geonameId":3571154,"toponymName":"Abondance","countryId":"3570311","fcl":"P","population":0,"countryCode":"MQ","name":"Abondance","fclName":"city, village,...","countryName":"Martinique","fcodeName":"populated place","adminName1":"Martinique","lat":"14.53733","fcode":"PPL"},{"adminCode1":"RE","lng":"55.67338","geonameId":936193,"toponymName":"Abondance","countryId":"935317","fcl":"P","population":0,"countryCode":"RE","name":"Abondance","fclName":"city, village,...","countryName":"Réunion","fcodeName":"populated place","adminName1":"Réunion","lat":"-21.03465","fcode":"PPL"}]}

> You will have convert the response similar to what is done for the simple search: https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/search/simplesearch.js#L299

OK thanks for the tip!

asaunier commented:
> CreateGeoJSONBloodhound expects GeoJSON. You will have convert the response similar to what is done for the simple search

Does this mean that I cannot use CreateGeoJSONBloodhound here since the response is not a GeoJSON? The ``filter`` function seems to be called on an already geojson-formatted response:
https://github.com/camptocamp/ngeo/blob/master/src/services/creategeojsonbloodhound.js#L78

Should I use a standard ``Bloodhound`` instead, as in https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/js/search/simplesearch.js#L271 ?

asaunier commented:
Other options:
* create a proxy service that will reformate the geonames reponses to GeoJSON
* use another service with a GeoJSON response. Perhaps Bing or Google geocoding services, but it seems both return JSON, not GeoJSON. https://developers.google.com/maps/documentation/geocoding/start https://msdn.microsoft.com/en-us/library/ff701711.aspx

tsauerwein commented:
We don't have to use `CreateGeoJSONBloodhound`.

asaunier commented:
> We don't have to use CreateGeoJSONBloodhound.

Yes, I have removed it and used a standard ``Bloodhound`` object.
The PR is almost ready:
* it seems there is a styling conflict with the simple search tool => the results are shown with a not really adapted style. In addition the country name is not visible (should be for instance ``Abondance (France)``). See
![sans titre](https://cloud.githubusercontent.com/assets/1192331/16498057/39c4b3fa-3efa-11e6-97e3-9eff44e507b7.png)
* when selecting a result, the recentering is not done to the correct place. There is probably a reprojection problem around here https://github.com/c2corg/v6_ui/blob/4a36a5d2fafd6ebb2270d069804155f03bf11279/c2corg_ui/static/js/map/search.js#L150-L155

Unfortunately I don't have time to fix the remaining problems. Feel free to fix them (especially the recentering issue) and to merge the PR, thanks :)

asaunier commented:
The PR may be tested in my dev instance at http://c2corgv6-demo.gis.internal:6555/waypoints

asaunier commented:
I have fixed the recentering problem and improved the layout of the suggestions dropdown even though it still needs to be fixed:
![sans titre](https://cloud.githubusercontent.com/assets/1192331/16947245/3a1f6c74-4dae-11e6-910e-cd526baafb0b.png)

@ginold I need your help because I am lost among all those typeahead styles!

May be tested live at http://c2corgv6-demo.gis.internal:6555/routes

asaunier commented:
Thanks @ginold for the layout improvement, it looks really good. Just one thing: do you know how we can get rid of the white left margin when the focus is on a result (blue line)?
![sans titre](https://cloud.githubusercontent.com/assets/1192331/16983125/aaaf895e-4e73-11e6-9801-2db166d5c48f.png)

asaunier commented:
> do you know how we can get rid of the white left margin when the focus is on a result (blue line)?

OK I found it. Will propose a small change...

asaunier commented:
> Will propose a small change...

I have removed ``margin-left: 5px;`` in https://github.com/c2corg/v6_ui/blob/map-search/less/map.less#L40 and it fixes the problem.

I have also rebased the PR.

@ginold OK for merging?

ginold commented:
Maybe the last thing I would suggest is to change the color while a suggestion is selected. Orange on blue background is bof. 

When moving with arrows, the suggestion div will have a `tt-cursor` (like `<div class="tt-suggestion tt-cursor">`) class, but you have to add the same styling when hovering.

At the moment I see it should be done for every app-simple search too (it has only been done on hover, not when moving with arrows), so you can add this style to simplesearch.less.

asaunier commented:
In my opinion the current styling is fair enough.
If you think it can be improved easily and without spending too much time, feel free to do it :)
